### PR TITLE
Akka.Persistence: improve `AsyncWriteJournal` and `PersistentActor` performance

### LINKED ISSUE #5522: [PERF] Reduce Akka.Persistence.Journal allocations + overhead
**Version Information**
Version of Akka.NET? 1.4.32
Which Akka.NET Modules? Akka.Persistence

**Describe the performance issue**
The current `AsyncWriteJournal` implementation needs some house cleaning in order to reduce excess allocations in the form of:

1. Boxing
2. Unnecessary array / collection copying
3. Delegate allocation
4. Possibly, inefficient LINQ usage (i.e. snapshotting with `ToList()` calls)

**Data and Specs**
Some examples pointed out by @to11mtm 

https://github.com/akkadotnet/akka.net/pull/5505#discussion_r783996956 
https://github.com/akkadotnet/akka.net/pull/5505#discussion_r783998326
https://github.com/akkadotnet/akka.net/pull/5505#discussion_r784008777

**Additional context**
Can use https://github.com/akkadotnet/akka.net/blob/dev/src/benchmark/Akka.Cluster.Benchmarks/Persistence/JournalWriteBenchmarks.cs to measure the improvements

### PR DESCRIPTION
**Pull request by Aaronontheweb**
## Changes

Fixes #5522

@F0b0s has already eliminated a ton of allocations from the journal - I'm addressing some more based on @to11mtm 's feedback. 

## Checklist

For significant changes, please ensure that the following have been completed (delete if not relevant):

* [x] This change follows the [Akka.NET API Compatibility Guidelines](https://getakka.net/community/contributing/api-changes-compatibility.html).
* [x] This change follows the [Akka.NET Wire Compatibility Guidelines](https://getakka.net/community/contributing/wire-compatibility.html).
* [x] I have [reviewed my own pull request](https://getakka.net/community/contributing/index.html#review-your-own-pull-requests).
* [x] Design discussion issue #5522

### Latest `dev` Benchmarks 

Include data from the [relevant benchmark](https://getakka.net/community/contributing/index.html#improve-performance) prior to this change here.

### This PR's Benchmarks

Include data from after this change here.

### CODE REVIEW

**Comment by Aaronontheweb** [comment]
# `JournalWriteBenchmarks`

## `dev` Benchmarks

``` ini

BenchmarkDotNet=v0.13.2, OS=Windows 10 (10.0.19044.2604/21H2/November2021Update)
AMD Ryzen 7 1700, 1 CPU, 16 logical and 8 physical cores
.NET SDK=7.0.100
  [Host]     : .NET 7.0.0 (7.0.22.51805), X64 RyuJIT AVX2
  Job-RHZXXN : .NET 7.0.0 (7.0.22.51805), X64 RyuJIT AVX2

InvocationCount=1  UnrollFactor=1  

```
|             Method | PersistentActors | WriteMsgCount |        Mean |     Error |     StdDev |      Median |      Gen0 |      Gen1 |   Allocated |
|------------------- |----------------- |-------------- |------------:|----------:|-----------:|------------:|----------:|----------:|------------:|
| **WriteToPersistence** |                **1** |           **100** |    **489.1 μs** |  **44.12 μs** |   **128.7 μs** |    **508.1 μs** |         **-** |         **-** |   **145.61 KB** |
| **WriteToPersistence** |               **10** |           **100** |  **1,289.8 μs** |  **62.97 μs** |   **175.5 μs** |  **1,279.5 μs** |         **-** |         **-** |  **1416.34 KB** |
| **WriteToPersistence** |              **100** |           **100** | **30,953.0 μs** | **943.54 μs** | **2,722.3 μs** | **30,163.0 μs** | **2000.0000** | **1000.0000** | **13980.24 KB** |


## This PR

``` ini

BenchmarkDotNet=v0.13.2, OS=Windows 10 (10.0.19044.2604/21H2/November2021Update)
AMD Ryzen 7 1700, 1 CPU, 16 logical and 8 physical cores
.NET SDK=7.0.100
  [Host]     : .NET 7.0.0 (7.0.22.51805), X64 RyuJIT AVX2
  Job-SDRCEZ : .NET 7.0.0 (7.0.22.51805), X64 RyuJIT AVX2

InvocationCount=1  UnrollFactor=1  

```
|             Method | PersistentActors | WriteMsgCount |        Mean |     Error |     StdDev |      Median |      Gen0 |      Gen1 |   Allocated |
|------------------- |----------------- |-------------- |------------:|----------:|-----------:|------------:|----------:|----------:|------------:|
| **WriteToPersistence** |                **1** |           **100** |    **521.7 μs** |  **45.74 μs** |   **132.0 μs** |    **535.2 μs** |         **-** |         **-** |   **147.91 KB** |
| **WriteToPersistence** |               **10** |           **100** |  **1,754.9 μs** | **210.71 μs** |   **611.3 μs** |  **1,580.8 μs** |         **-** |         **-** |  **1411.19 KB** |
| **WriteToPersistence** |              **100** |           **100** | **29,833.8 μs** | **725.73 μs** | **2,093.9 μs** | **29,651.6 μs** | **2000.0000** | **1000.0000** | **13893.84 KB** |

**Comment by Aaronontheweb** [comment]
Got rid of the value delegates - that seemed to hurt more than help.

``` ini

BenchmarkDotNet=v0.13.2, OS=Windows 10 (10.0.19044.2604/21H2/November2021Update)
AMD Ryzen 7 1700, 1 CPU, 16 logical and 8 physical cores
.NET SDK=7.0.100
  [Host]     : .NET 7.0.0 (7.0.22.51805), X64 RyuJIT AVX2
  Job-PWUWVX : .NET 7.0.0 (7.0.22.51805), X64 RyuJIT AVX2

InvocationCount=1  UnrollFactor=1  

```
|             Method | PersistentActors | WriteMsgCount |        Mean |     Error |     StdDev |      Median |      Gen0 |      Gen1 |   Allocated |
|------------------- |----------------- |-------------- |------------:|----------:|-----------:|------------:|----------:|----------:|------------:|
| **WriteToPersistence** |                **1** |           **100** |    **444.4 μs** |  **46.21 μs** |   **134.8 μs** |    **443.2 μs** |         **-** |         **-** |   **148.09 KB** |
| **WriteToPersistence** |               **10** |           **100** |  **1,322.2 μs** | **102.98 μs** |   **290.5 μs** |  **1,197.5 μs** |         **-** |         **-** |  **1413.47 KB** |
| **WriteToPersistence** |              **100** |           **100** | **30,274.7 μs** | **732.28 μs** | **2,041.3 μs** | **29,870.2 μs** | **2000.0000** | **1000.0000** | **13956.02 KB** |

**Comment by Aaronontheweb** [comment]
Eliminated some additional `.ToArray` calls that should be unnecessary (unless the Akka.Persistence API contract is wrong, in which case we should require something more robust like `IReadOnlyList<T>`.)

``` ini

BenchmarkDotNet=v0.13.2, OS=Windows 10 (10.0.19044.2604/21H2/November2021Update)
AMD Ryzen 7 1700, 1 CPU, 16 logical and 8 physical cores
.NET SDK=7.0.100
  [Host]     : .NET 7.0.0 (7.0.22.51805), X64 RyuJIT AVX2
  Job-AKPKTM : .NET 7.0.0 (7.0.22.51805), X64 RyuJIT AVX2

InvocationCount=1  UnrollFactor=1  

```
|             Method | PersistentActors | WriteMsgCount |        Mean |     Error |     StdDev |      Median |      Gen0 |      Gen1 |   Allocated |
|------------------- |----------------- |-------------- |------------:|----------:|-----------:|------------:|----------:|----------:|------------:|
| **WriteToPersistence** |                **1** |           **100** |    **449.6 μs** |  **48.75 μs** |   **143.0 μs** |    **441.4 μs** |         **-** |         **-** |   **145.92 KB** |
| **WriteToPersistence** |               **10** |           **100** |  **1,262.3 μs** |  **75.83 μs** |   **207.6 μs** |  **1,172.6 μs** |         **-** |         **-** |   **1383.5 KB** |
| **WriteToPersistence** |              **100** |           **100** | **29,764.8 μs** | **769.96 μs** | **2,221.5 μs** | **29,278.7 μs** | **2000.0000** | **1000.0000** | **13679.48 KB** |

File: src/contrib/cluster/Akka.Cluster.Sharding.Tests.MultiNode/AsyncWriteProxyEx.cs [hunk 1]
```diff
@@ -162,26 +168,25 @@ protected override bool AroundReceive(Receive receive, object message)
         /// <returns>TBD</returns>
         protected override Task<IImmutableList<Exception>> WriteMessagesAsync(IEnumerable<AtomicWrite> messages)
         {
+            var trueMsgs = messages.ToArray();
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.Cluster.Sharding.Tests.MultiNode/AsyncWriteProxyEx.cs:171]: work-around for serialization problems caused by the MNTR Cluster.Sharding specs that use this journal impl.

**Review by Arkatufus — APPROVED** [review]
LGTM

**Pull request merged**